### ISSUE TITLE
Make Dummy ValueTuples' members Dummies, to match Tuple

### ISSUE THREAD

**Issue by TimLovellSmith** [FakeItEasy/FakeItEasy#1637]
Repreod in 5.2 also.

I got bitten by unexpected behavior with FakeItEasy 2.3.0 (before I updated) while I was doing some 'refactoring'. I was trying to take advantage of the new tuple language features. Unfortunately I found this not-quite-innocent-as-it-looks change actually broke my tests.

`ArmCreatableResourceFactory` below is a fake that had no behavior configured for `CreateStorageAccountAsync`, so it got the default call behavior...

WORKS: 

```c#
Tuple<StorageAccount, string> storageAccountAndId = await ArmCreatableResourceFactory.CreateStorageAccountAsync(dep, scope, resourceGroup, token, ExecCtx);
storageAccount = storageAccountAndId.Item1;
string storageAccountId = storageAccountAndId.Item2;
var x = storageAccount.AccountName;
```

DOESN'T WORK:
```c#
var (storageAccount, storageAccountId) = await ArmCreatableResourceFactory.CreateStorageAccountAsync(dep, scope, resourceGroup, token, ExecCtx);
var x = storageAccount.AccountName;
```
                
The reason it doesn't work is while the code runs, it results in different state, and when you dereference `storageAccount.AccountName`, in the second instance it is null!

The main impact of the issue is this kind of brittleness makes the code harder to refactor.

**Comment by blairconrad** [comment]
Sorry you're having problems, @TimLovellSmith.

I've not kept up on the `ValueTuple` bandwagon. What's the signature of `CreateStorageAccountAsync`? Did it change between WORKS and DOESN'T WORK?

**Comment by blairconrad** [comment]
I made a probably very poor attempt to reproduce, but this all passed:

```c#
public class Source
{
    public virtual async Task<Tuple<int, string>> Make() => Tuple.Create(3, "d");
}
class Program
{
    static async Task Main(string[] args)
    {
        var fake1 = A.Fake<Source>();
        Tuple<int, string> result1 = await fake1.Make();
        var fake2 = A.Fake<Source>();
        var (id, name) = await fake2.Make();
        result1.Item2.Should().Be("");
        name.Should().Be("");
    }
}
```

**Comment by thomaslevesque** [comment]
@blairconrad in this code you're not using `ValueTuple` at all, you're just using tuple deconstruction.
`var (id, name) = await fake2.Make()` is equivalent to this:

```
Tuple<int, string> tmp =  await fake2.Make();
tmp.Deconstruct(out int id, out string name);
```

(`Deconstruct` is [an extension method](https://docs.microsoft.com/en-us/dotnet/api/system.tupleextensions.deconstruct?view=netframework-4.8#System_TupleExtensions_Deconstruct__2_System_Tuple___0___1____0____1__) on `Tuple<T1, T2>`)

@TimLovellSmith I'm not sure I really understand what your problem is. A complete example would help, as well as the answer to this question:

> What's the signature of `CreateStorageAccountAsync`? Did it change between WORKS and DOESN'T WORK?

If you didn't change the signature, but only the call site, the problem is probably not related to FakeItEasy.

**Comment by thomaslevesque** [comment]
OK, I repro'd the issue.

```csharp

void Main()
{
    var foo = A.Fake<IFoo>();
    
    {
        var accountAndId = foo.GetAccountAndIdTuple();
        string accountName = accountAndId.Item1.AccountName;
        Console.WriteLine(accountName == null ? "null" : accountName.Length > 0 ? accountName : "(empty)");
    }
    
    {
        var (account, id) = foo.GetAccountAndIdValueTuple();
        string accountName = account.AccountName; // NullReferenceException here, account is null
        Console.WriteLine(accountName == null ? "null" : accountName.Length > 0 ? accountName : "(empty)");
    }
}

public interface IFoo
{
    Tuple<StorageAccount, string> GetAccountAndIdTuple();
    (StorageAccount, string) GetAccountAndIdValueTuple();
}

public class StorageAccount
{
    public StorageAccount(string accountName)
    {
        AccountName = accountName;
    }
    
    public string AccountName { get; }
}
```

I think this is related to how we create dummies. `ValueTuple<>` is a value type, so we use the `ResolveByActivatingValueTypeStrategy` strategy, which creates an instance of the type with the default constructor (which is equivalent to returning `default(T)`). Unlike reference types, we never try to call a constructor with parameters.

I see two main options to fix this:
- Change how we resolve dummies for non-primitive value types, to do the same as for reference types (i.e. using a constructor with arguments, resolving the arguments as dummies). It might be the right thing to do, but it could have side effects, and possibly break existing code.
- Special-case `ValueTuple`. This type is going to be more and more widely used, so it might be worth it. The problem is that `ValueTuple` is not in `net4x` and `netstandard1.x`, so we'd have to either reference the package that contains it, or handle it using reflection.

In the meantime, @TimLovellSmith, you can create a [dummy factory](https://fakeiteasy.readthedocs.io/en/stable/custom-dummy-creation/) to tell FakeItEasy how to create instances of `ValueTuple<>`.

Something like this:

```csharp
public class ValueTupleDummyFactory : IDummyFactory
{
    public Priority Priority => Priority.Default;

    public bool CanCreate(Type type)
    {
        return type.IsGenericType
            && !type.IsGenericTypeDefinition
            && type.FullName.StartsWith("System.ValueTuple`");
    }

    public object Create(Type type)
    {
        // Create dummies for each member of the tuple
        var args = Array.ConvertAll(type.GetGenericArguments(), FakeItEasy.Sdk.Create.Dummy);
        return Activator.CreateInstance(type, args);
    }
}
```

**Comment by thomaslevesque** [comment]
Writing that dummy factory makes me realize that special-casing `ValueTuple` is pretty easy, even without referencing the nuget package, so that's probably the least risky approach.

We can still reconsider how we resolve dummies of value types at a later time if necessary.

**Comment by blairconrad** [comment]
Bah. You beat me to it. On the drive into work, I had the same realization.
Me, I think we should patch now, but I like expanding support for non-primitive value types. As you say, it might be breaking, so how about we create an issue and schedule it into the next major release?

**Comment by thomaslevesque** [comment]
> Me, I think we should patch now

Will do. I suggest we just add a built-in dummy factory, like we did for `String`.

> but I like expanding support for non-primitive value types. As you say, it might be breaking, so how about we create an issue and schedule it into the next major release?

Sounds good. I'll create a new issue.

**Comment by TimLovellSmith** [comment]
Sorry for the incomplete repro details! Yes, Thomas figured out exactly the issue I was having. Yes, I think special-casing ValueTuple is probably the way to go... That might raises a new question, of whether object 'Tuple' should be handled identically to ValueTuple. 

IMHO, there is little utility in creating Tuple dummy types - Tuples should really just always behave like mathematical 'tuples', where they are seen as a bunch of values grouped together. So you shouldn't be trying to configure them. But I worry that this would unfortunately be a breaking change...

**Comment by thomaslevesque** [comment]
> Yes, I think special-casing ValueTuple is probably the way to go... That might raises a new question, of whether object 'Tuple' should be handled identically to ValueTuple.

That's the plan. `Tuple<>` doesn't need any special handling, we treat it as any other class, by passing dummy values to its constructor. We're going to do the same for `ValueTuple<>` (except that it *will* need special handling).

> IMHO, there is little utility in creating Tuple dummy types - Tuples should really just always behave like mathematical 'tuples', where they are seen as a bunch of values grouped together. So you shouldn't be trying to configure them. But I worry that this would unfortunately be a breaking change...

Dummies are not configurable (unless when they happen to be fakes, but that's accidental, not intentional).
In the case of `Tuple<T>`, the class isn't sealed, so we actually create a fake when we need a dummy. But since the interesting members (`Item1` and `Item2`) are not virtual, there isn't much to configure anyway...

**Comment by TimLovellSmith** [comment]
>In the case of Tuple<T>, the class isn't sealed, so we actually create a fake when we need a dummy. But since the interesting members (Item1 and Item2) are not virtual, there isn't much to configure anyway...

I don't mean anybody should *want* to configure them. I just wasn't sure if it were previously possible. It sounds like it is not actually possible. So since Tuple has nothing fakeable or dummyable, wouldn't it be simpler to explain in your docs if Tuple and ValueTuple worked the same way and just returned plain old Tuple/ValueTuple object (with fake/dummy fields) and not a Fake/Dummy object?
(And may even hypothetically be a perf win for some artificial benchmarks... - but likely not every real world scenario.)

I could extend this same argument for some other non-fakable types too, like arrays. The tricky bit is knowing when to stop arguing that yet one more object type doesn't need to be faked/dummied...

**Comment by blairconrad** [comment]
I'm not entirely sure what your goal is, @TimLovellSmith. A [Dummy](https://fakeiteasy.readthedocs.io/en/stable/dummies/) isn't necessarily a special kind of object. It's really only special in how it's created and that it's guaranteed not to be null. Some Dummies _are_ Fakes, but many are not. In the `Tuple` case, we do return a Fake object as the Dummy, but that's because it's a non-sealed class that's got an accessible constructor. Any class meeting those requirements is Faked in the same way when a Dummy is requested. We could special case the class and provide a non-Fake, explaining so in the docs, but in some sense that's more work. Certainly for us, tracking down all those special cases.

If you're just suggesting we make changes for performance reasons, I agree with you that it's probably not worth it. There's a small optimization for `Tuple`, but as @thomaslevesque says, a Dummy that pops out is effectively the same as a real `Tuple`, so any customization would just be something for us to maintain.

And like you say, an array is not fakeable, so we create an actual 0-element array, so there's not much to optimize, and probably no reason to mention it specifically in the docs…

**Comment by blairconrad** [comment]
@TimLovellSmith, we released 5.3.0-beta.1 with this fix. Are you interested in trying it out before we release 5.3.0?

**Comment by TimLovellSmith** [comment]
Sounds good to me!

**Comment by blairconrad** [comment]
Thanks, @TimLovellSmith. We await your feedback!

**Comment by TimLovellSmith** [comment]
@blairconrad  So.. this seems to work great. The really surprising bit, unless if I am not getting confused (as I oft do), is that my tests seem to run about 10% faster with this change than I remember them running last week with 5.2 and 2.3. I am really not sure why, or how reproducible that might be.

**Comment by thomaslevesque** [comment]
Hi @TimLovellSmith. I don't think the changes we made have much effect on performance, so it might be accidental (e.g. less load on your machine). Or maybe your tests use a lot of tuples, and we can now make them faster ;)

**Comment by blairconrad** [comment]
I'm glad it works, @TimLovellSmith. Like @thomaslevesque says, there doesn't appear to be anything in the latest release that would improve performance much, unless you use _a lot_ of tuples. Like just an astounding amount.
However, we did a lot of performance-related work in the [5.0.0 release](https://github.com/FakeItEasy/FakeItEasy/releases/tag/5.0.0), which _might_ account for some of the difference relative 2.3.0…

**Comment by TimLovellSmith** [comment]
Could maybe just be the savings of not having to take locks to generate types for Tuple<string,int>, Tuple<int,int>, Tuple<class1,string>, Tuple<IPAddress,int>, etc, etc, etc... too many tuples! 
This brings me to yet another idea...

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 5.3.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/5.3.0).

Thanks, @TimLovellSmith. Look for your name in the release notes! 🏆
